surefire-providers/surefire-junit-platform/src/main/java/org/apache/maven/surefire/junitplatform/RunListenerAdapter.java

```diff
@@ -216,11 +217,25 @@ private SimpleReportEntry createReportEntry(
             String reason,
             Integer elapsedTime) {
         String[] classMethodName = toClassMethodName(testIdentifier);
-        String className = classMethodName[0];
+
+        Optional<UniqueId.Segment> classSegment = testIdentifier.getUniqueIdObject().getSegments().stream()
+                .filter(segment -> "class".equals(segment.getType()))
+                .findFirst();
```

Comment:
You shouldn't parse the `UniqueId` because it's considered to be an implementation detail. Instead, you should use `testPlan.getParent(testIdentifier)` to find the right level. If this should only support class-based engines, finding the last one with a `TestSource` of type `ClassSource` should work. If this should also work for file-based engines like Cucumber, you might want to use a differeny heuristic instead. Maybe always use the level immediate below the root (engine descriptor)? If it has a `ClassSource`, use that; otherwise, use the display name (similar to what `toClassMethodName` does)?

Reply:
Yeah, Cucumber was also [broken by the last patch](https://issues.apache.org/jira/browse/SUREFIRE-2299), so the alternative heuristic should be used.

Reply:
> You shouldn't parse the `UniqueId` because it's considered to be an implementation detail. Instead, you should use `testPlan.getParent(testIdentifier)` to find the right level. If this should only support class-based engines, finding the last one with a `TestSource` of type `ClassSource` should work. If this should also work for file-based engines like Cucumber, you might want to use a differeny heuristic instead. Maybe always use the level immediate below the root (engine descriptor)? If it has a `ClassSource`, use that; otherwise, use the display name (similar to what `toClassMethodName` does)?

I tried this approach but it looks not possible with the `Expected reporting` from here https://github.com/apache/maven-surefire/issues/2601#issue-3201332136